File: sdk/java/src/main/java/com/ouc/tcp/sdk/util/Checksum.java
package com.ouc.tcp.sdk.util;

public final class Checksum {
    private Checksum() {}

    public static int internetChecksum(byte[] data) {
        int length = data.length;
        int i = 0;
        long sum = 0;

        while (length > 1) {
            int value = ((data[i] << 8) & 0xFF00) | (data[i + 1] & 0xFF);
            sum += value & 0xFFFF;
            if ((sum & 0xFFFF0000) != 0) {
                sum = (sum & 0xFFFF) + (sum >> 16);
            }
            i += 2;
            length -= 2;
        }

        if (length > 0) {
            sum += (data[i] << 8) & 0xFF00;
            if ((sum & 0xFFFF0000) != 0) {
                sum = (sum & 0xFFFF) + (sum >> 16);
            }
        }

        return (int) (~sum) & 0xFFFF;
    }
}

File: sdk/rust/src/checksum.rs
/// Simple 16-bit Internet checksum (ones' complement) utility.
/// This helper is optional for RDT1 but becomes useful once students
/// implement error detection in RDT2+.
pub fn internet_checksum(data: &[u8]) -> u16 {
    let mut sum: u32 = 0;
    let mut chunks = data.chunks_exact(2);

    for chunk in &mut chunks {
        let value = u16::from_be_bytes([chunk[0], chunk[1]]) as u32;
        sum = sum.wrapping_add(value);
    }

    if let Some(&byte) = chunks.remainder().first() {
        sum = sum.wrapping_add((byte as u32) << 8);
    }

    while (sum >> 16) != 0 {
        sum = (sum & 0xFFFF) + (sum >> 16);
    }

    !(sum as u16)
}
